vllm/model_executor/models/transformers.py

```diff
@@ -0,0 +1,257 @@
+    return attention_instances[layer_idx].forward(
+        query,
+        key,
+        value,
+        _kv_cache=kv_caches[layer_idx],
+        _attn_metadata=attn_metadata), None
+
+
+ALL_ATTENTION_FUNCTIONS["vllm"] = vllm_flash_attention_forward
+
+
+# Linear Layer that is compatible with transformers internal forward
+# TODO: This is a temporary solution, we should find a better way to integrate
+class HFColumnParallelLinear(ColumnParallelLinear):
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return super().forward(input)[0]
+
+
+class HFRowParallelLinear(RowParallelLinear):
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return super().forward(input)[0]
+
+
+def replace_tp_linear_class(orig_module: nn.Linear, style: str):
+    """
+    In model configurations, we use a neutral type (string) to specify parallel
+    styles, here we use it to translate nn.Linear into vllm-style tp Linear.
+    """
+
+    if not isinstance(style, str):
+        raise ValueError(
+            f"Unsupported parallel style type {type(style)}, expected str")
+
+    input_size = orig_module.in_features
+    output_size = orig_module.out_features
+    bias = orig_module.bias is not None
+
+    if style == "colwise":
+        return HFColumnParallelLinear(input_size, output_size, bias)
+    elif style == "rowwise":
+        return HFRowParallelLinear(input_size, output_size, bias)
+    # We don't consider colwise_rep since it's used in lm_head
+    else:
+        raise ValueError(f"Unsupported parallel style value: {style}")
+
+
+class TransformersModel(nn.Module, SupportsLoRA):
+    embedding_padding_modules = ["lm_head"]
+
+    # TODO Add support for bnb and LORA
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = "") -> None:
+        super().__init__()
+        logger.info("Using Transformers backend.")
+
+        self.vllm_config = vllm_config
+        config = vllm_config.model_config.hf_config
+        cache_config = vllm_config.cache_config
+        quant_config = vllm_config.quant_config
+        self.config = config
+        self.vocab_size = config.vocab_size
+        self.unpadded_vocab_size = config.vocab_size
+
+        self.model: PreTrainedModel = AutoModel.from_config(
+            self.config,
+            torch_dtype=vllm_config.model_config.dtype,
+            trust_remote_code=vllm_config.model_config.trust_remote_code,
+        )
+        prefix = self.model.base_model_prefix
+
+        # MLP modifications
+        self.tensor_parallelize(self.model)
+
+        # Attention modifications (assumes 1 attention op per hidden layer)
+        tp_size = get_tensor_model_parallel_world_size()
+        self.attention_instances = [
+            Attention(
+                divide(config.num_attention_heads, tp_size),
+                config.head_dim,
+                config.head_dim**
+                -0.5,  # ish, the scaling is different for every attn layer
```

Comment:
Could you name this variable and give a larger NOTE or TODO on it? I assume this will not be sufficient for all models

Reply:
I've updated this to explain what we're doing, and also added an override which should apply the correct scale for models that don't use Llama's value